pkg/registry/core/service/storage/rest.go

```diff
@@ -130,6 +139,53 @@ func (rs *REST) Categories() []string {
 	return []string{"all"}
 }
 
+func (rs *REST) getNetworkStorage(tenant string) (rest.Getter, error) {
+	if atomic.LoadUint32(&rs.done) == 1 {
+		if rs.networks == nil {
+			return nil, fmt.Errorf("service storage failed to identify proper network getter")
+		}
+		return rs.networks, nil
+	}
+
+	// Slow-path.
+	rs.m.Lock()
+	defer rs.m.Unlock()
+	if rs.done == 0 {
+		if rs.networks != nil {
+			defer atomic.StoreUint32(&rs.done, 1)
+			return rs.networks, nil
+		}
+
+		crStorageGetter := extensionsapiserver.GetCustomResourceStoragesGetter()
+		if crStorageGetter == nil {
+			defer atomic.StoreUint32(&rs.done, 1)
+			return nil, fmt.Errorf("api extension server not initialized properly")
+		}
+
+		if storage, err := crStorageGetter.GetCustomResourceStorage(tenant, "networks.arktos.futurewei.com", "v1"); err == nil {
+			defer atomic.StoreUint32(&rs.done, 1)
+			rs.networks = storage.CustomResource
+			return rs.networks, nil
+		} else {
+			return nil, fmt.Errorf("custom resource storage not set up yet")
+		}
+	}
+
+	if rs.networks == nil {
```

Comment:
I am just curious, when the code will hit line 174? When will rs.done == 1 && atomic.loadUnit32(rs.done) == 0?

Reply:
see sync.Once code; if this method is called in different goroutines simultaneously, it is possible to reach this line.

Reply:
Thanks for sharing the knowledge.